### PR TITLE
bundle: properly generate resources for custom OPERATOR_NAMEPREFIX

### PR DESCRIPTION
**Pull request by jarrpa**
Some of the CSI RBAC manifests were not formatted quite like kustomize expects, apparently. The following additions ensure all RBAC resources are created with appropriate references to each other in the correct namespaces (where applicable). This necessitated changes to address various issues resulting from people doing builds with a value for `OPERATOR_NAMEPREFIX` different from our default.

### CODE REVIEW

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
Hey @jarrpa thanks for the cleanup 

I tried to deploy the changes in a new ocp cluster with namespace `ocs-test` but the clusterrolebinding/rolebinding are pointing to the `ocs-client-operator-system` which is not expected. can you please check required changes at https://github.com/red-hat-storage/ocs-client-operator/blob/ed42874cb82f2a9645ea699b3f61b44b207a0f16/Makefile#L111

**Comment by jarrpa** [comment]
@Madhu-1 I was trying to find an example of how to change that within kustomize, the fact that we'd have to use `sed` just seems so... hackish. But for lack of anything else, is there an example of someone already doing this?

**Comment by nbalacha** [comment]
> @Madhu-1 I was trying to find an example of how to change that within kustomize, the fact that we'd have to use `sed` just seems so... hackish. But for lack of anything else, is there an example of someone already doing this?

@iamniting figured out a way to do this for the LVM and OCS operators . He can explain but 

> @Madhu-1 I was trying to find an example of how to change that within kustomize, the fact that we'd have to use `sed` just seems so... hackish. But for lack of anything else, is there an example of someone already doing this?

@iamniting figured out how to do this for the lvm and ocs operators . I don't think it requires a sed.
https://github.com/openshift/lvm-operator/pull/298/files

**Comment by iamniting** [comment]
> > @Madhu-1 I was trying to find an example of how to change that within kustomize, the fact that we'd have to use `sed` just seems so... hackish. But for lack of anything else, is there an example of someone already doing this?
> 
> @iamniting figured out a way to do this for the LVM and OCS operators . He can explain but
> 
> > @Madhu-1 I was trying to find an example of how to change that within kustomize, the fact that we'd have to use `sed` just seems so... hackish. But for lack of anything else, is there an example of someone already doing this?
> 
> @iamniting figured out how to do this for the lvm and ocs operators . I don't think it requires a sed. https://github.com/openshift/lvm-operator/pull/298/files

@nbalacha It should not have been done to hardcode the name "ocs-client-operator" at all the locations in the rbacs; it will be done automatically by kustomize using a prefix, therefore I believe @jarrpa is trying to update the names.

Also, i don't see bundle changes in the PR, @jarrpa I suggest you guys use `--extra-service-accounts` flag while generating the bundle if all the permissions are not related to the operator only. Using `--extra-service-accounts` flag will move the rbacs in the CSV which will help you to install the operator in any namespace.

**Comment by jarrpa** [comment]
> Also, i don't see bundle changes in the PR, @jarrpa I suggest you guys use `--extra-service-accounts` flag while generating the bundle if all the permissions are not related to the operator only. Using `--extra-service-accounts` flag will move the rbacs in the CSV which will help you to install the operator in any namespace.

Already done: https://github.com/red-hat-storage/ocs-client-operator/blob/main/Makefile#L111 And we seem to already have them in the CSV as well, these changes are for development deployment without OLM. Though now that you mention it we should probably also remove the extra manifests from the bundle directory.

**Comment by jarrpa** [comment]
@Madhu-1 I think I got it. By specifying the namespace and namespace prefix as separate we can pass them to kustomize via `edit set`.

**Review by Madhu-1 — APPROVED** [review]
Changes LGTM if deployment works in some custom namespaces.

**Comment by Madhu-1** [comment]
```
[🎩︎]mrajanna@fedora ocs-client-operator $]oc get rolebinding ocs-client-operator-test-csi-rbd-plugin-rb -oyaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  creationTimestamp: "2023-04-04T07:30:38Z"
  name: ocs-client-operator-test-csi-rbd-plugin-rb
  namespace: ocs-client-ns-test
  resourceVersion: "10105"
  uid: 55d12629-f235-40c5-a48e-7b638069a8cd
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: ocs-client-operator-test-csi-rbd-plugin-role
subjects:
- kind: ServiceAccount
  name: ocs-client-operator-test-csi-rbd-plugin-sa
  namespace: ocs-client-operator-test-system
```
```
[🎩︎]mrajanna@fedora ocs-client-operator $]oc get po ocs-client-operator-test-controller-manager-8498956f46-ps6xx -oyaml |grep -i namespace
    olm.operatorNamespace: ocs-client-ns-test
    olm.targetNamespaces: ocs-client-ns-test
  namespace: ocs-client-ns-test
    - name: OPERATOR_NAMESPACE
          fieldPath: metadata.namespace
              fieldPath: metadata.namespace
            path: namespace
```

With this PR, the SA namespace in role binding is pointed to the wrong namespace `ocs-client-operator-test-system`, which is unexpected. The role and role binding will get created on the namespace where the operator pod is running in this case, its `ocs-client-ns-test`

File: Makefile [hunk 1]
```diff
@@ -113,13 +113,19 @@ remove-with-olm: ## Remove controller from the K8s cluster
...
 .PHONY: bundle
 bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metadata, then validate generated files.
+	rm -rf ./bundle
 	$(OPERATOR_SDK) generate kustomize manifests -q
 	cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
-	cd config/default && $(KUSTOMIZE) edit set image rbac-proxy=$(RBAC_PROXY_IMG)
-	cd config/manifests/bases && $(KUSTOMIZE) edit add annotation --force 'olm.skipRange':"$(SKIP_RANGE)" && \
+	cd config/default && \
+		$(KUSTOMIZE) edit set image kube-rbac-proxy=$(RBAC_PROXY_IMG) && \
+		$(KUSTOMIZE) edit set namespace $(OPERATOR_NAMESPACE) && \
```
Comment (Madhu-1) [inline Makefile:121]: After removing this `set namespace` line from the Makefile i can deploy the operator in a different namespace. @jarrpa can you remove this line?

**Review by Madhu-1 — COMMENTED** [review]
yq is not getting installed can you please check

```bash
[🎩︎]mrajanna@fedora ocs-client-operator $]make bundle
Downloading sigs.k8s.io/controller-tools/cmd/controller-gen@v0.4.1
/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/controller-gen "crd:trivialVersions=true,preserveUnknownFields=false" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
Downloading sigs.k8s.io/kustomize/kustomize/v4@v4.5.3
Downloading https://github.com/operator-framework/operator-sdk/releases/download/v1.20.0/operator-sdk_linux_amd64
rm -rf ./bundle
/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/operator-sdk generate kustomize manifests -q
cd config/manager && /home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/kustomize edit set image controller=quay.io/ocs-dev/ocs-client-operator:latest
cd config/default && \
	/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/kustomize edit set image kube-rbac-proxy=registry.redhat.io/openshift4/ose-kube-rbac-proxy:v4.9.0 && \
	/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/kustomize edit set nameprefix ocs-client-operator-
cd config/manifests/bases && \
	/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/kustomize edit add annotation --force 'olm.skipRange':"" && \
	/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/kustomize edit add patch --name ocs-client-operator.v0.0.0 --kind ClusterServiceVersion\
	--patch '[{"op": "replace", "path": "/spec/replaces", "value": ""}]'
2023/04/04 11:54:56 patch types.Patch{Path:"", Patch:"[{\"op\": \"replace\", \"path\": \"/spec/replaces\", \"value\": \"\"}]", Target:(*types.Selector)(0xc00031dc80), Options:map[string]bool(nil)} already in kustomization file
/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/kustomize build config/manifests | sed "s|STATUS_REPORTER_IMAGE_VALUE|quay.io/ocs-dev/ocs-client-operator:latest|g" | awk '{print}'| \
	/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/operator-sdk generate bundle -q --overwrite --version 4.12.0 --channels=alpha --default-channel=alpha --extra-service-accounts="$(/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/kustomize build config/manifests | /home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/yq 'select(.kind == "ServiceAccount") | .metadata.name' -N | paste -sd "," -)"
bash: line 2: /home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/yq: No such file or directory
WARN[0000] ClusterServiceVersion validation: [OperationFailed] provided API should have an example annotation 
WARN[0000] ClusterServiceVersion validation: [OperationFailed] provided API should have an example annotation 
INFO[0000] Creating bundle.Dockerfile                   
INFO[0000] Creating bundle/metadata/annotations.yaml    
INFO[0000] Bundle metadata generated suceessfully       
sed -i "s|packageName:.*|packageName: csi-addons|g" "config/metadata/dependencies.yaml"
sed -i "s|version:.*|version: ""0.5.0""|g" "config/metadata/dependencies.yaml"
cp config/metadata/* bundle/metadata/
/home/mrajanna/workspace/src/github.com/red-hat-storage/ocs-client-operator/bin/operator-sdk bundle validate ./bundle
WARN[0000] Warning: Value ocs.openshift.io/v1alpha1, Kind=StorageClassClaim: provided API should have an example annotation 
WARN[0000] Warning: Value ocs.openshift.io/v1alpha1, Kind=StorageClient: provided API should have an example annotation 
INFO[0000] All validation tests have completed successfully 
```

File: Makefile (hunk 1 above)
Comment (Madhu-1) [reply Makefile:121]: Looks like something wrong with my cluster. everything works fine, sorry for the noise.

Comment (jarrpa) [reply Makefile:121]: Over a call, we discovered that something may have gone wrong in Madhu's environment. We verified on both our ends that, after making sure our shell environments were correct, the bundle installation worked correctly.

**Comment by jarrpa** [comment]
> yq is not getting installed can you please check

Valid, forgot to add the target prereq. Fixed!

**Pull request merged**